### ISSUE TITLE
Package submission request for waddR

### ISSUE THREAD

**Issue by JulianFlesch** [Bioconductor/Contributions#1218]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/goncalves-lab/waddR

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by JulianFlesch** [comment]
I will reopen this as a new issue, as this was created prematurely. A push to master was not yet considered in this request. Please excuse this mistake.

**Comment by JulianFlesch** [comment]
Thank you @mtmorgan, following your suggestion I am now reopening this issue.

**Comment by JulianFlesch** [comment]
It appears, the bioc-issue-bot has not automatically detected the new version after reopening this issue. I will now push a version bump to trigger the build.

**Comment by Liubuntu** [comment]
Hi @JulianFlesch , 

Please see here for the initial review. Your package looks good. Mostly of the reviews are about suggestions, not really issues. Please try to look through and implement as your interest. Comment back here with any questions, comments, updates and when you are ready for a 2nd look.

Cheers,
Qian

## DESCRIPTION
- Remove the `Date` field.
- `BugReports` better to use the issue tracker on github. https://github.com/goncalves-lab/waddR/issues

## NAMESPACE
- Looks good!

## R/

- RECOMMEND: Remove the prefix of `.` for exporting functions. The `.`
  perfix is conventionally used for unexported utility functions.
- Question: since you already have the `wasserstein.test()` to
  dispatch different method, is it still necessary to export the
  `wasserstein.test.sp()` and `wasserstein.test.asy()`? Given that the
  documentations are necessarily duplicated also... It's totally fine
  to keep all exported though.

`wasserstein.test()` function improvement suggestions (current version
is fine):

- RECOMMEND: put the `method` argument before `permnum`, so that a
  simple call of `wasserstein.test(x, y, "SP")` could be used without
  specifying the `method` argument when the default value of `permnum`
  is used.

- RECOMMEND: define default values for `method` argument. e.g.,
  ```
  function(x, y, method = c("sp", "ASY"), permnum = 10000) {
      method <- match.arg(method)
  ```

  This way, user could leave this argument blank, and the default
  behavior will use `sp` in `wasserstein.test(x, y)`. The
  `match.arg(method)` will check the user's input, and stop if doesn't
  match.

- RECOMMEND: To further simplify code, use `switch` to dispatch the
  `method` to different tests. e.g.,

  ```
  switch(method,
      "SP" = wasserstein.test.sp(x, y, permnum),
      "ASY" = wasserstein.test.asy(x, y))
  ```
  try `?match.arg` for examples.

## man/

- `wasserstein_metric()` DESCRIPTION, typo in `trnasport`.

## vignette
- Just a note, that the development version of package could also be
  installed from github directly using
  `BiocManager::install("githubUsername/pkgName")`, which is
  equivalent to `devtools::install_github("../..")`.
- Add `sessionInfo()` in all vignette files, because different set of
  packages were loaded.

## unit tests
- `test_wasserstein_metric.r`, remove commented code about `cat`s.

**Comment by JulianFlesch** [comment]
Hi @Liubuntu ,

Thank you for taking the time to review our package! In the above commits I addressed all of the points you recommended for change. 

## Most important changes to the R code:
* Changed the order of arguments and added defaults in `wasserstein.test` and `wasserstein.sc`
* Changed the method dispatch to use a switch statement in both functions
* Removed the functions `wasserstein.test.sp` and `wasserstein.test.asy` from the NAMESPACE in favor of the dispatch method `wasserstein.test`

## Changes to Vignettes:
* Added SessionInfo() to all vignettes
* Updated the install instructions
* Updated function calls to reflect the changes to `wasserstein.test` and `wasserstein.sc`

## Other:
* Fixed a bug
* Fixed typos
* Cleaned up a test
* Excluded functions from NAMESPACE

Thank you again for helping us!

Best,
Julian

**Comment by JulianFlesch** [comment]
## Question About a Reproducibility Feature

Hi again,

There is one more feature that was requested in our group with regard to the permutation procedure performed when calling `.wassersteinTestSp`. To calculate p-values in this method, the Wasserstein distance between random permutations of the two input vectors is calculated, which involves a random sampling (by calling the Rcpp Sugar function `sample` inside the function `permutate`).
It has been requested to use seeds in order to make the p-values reproducible in re-runs on the same data-sets.

It has been described how this can be achieved in Cpp by calling `set.seed` from R (see [this github issue](https://github.com/RcppCore/Rcpp/issues/794)), however, the use of `set.seed` seems to be discouraged by Bioconductor. In [this thread](https://support.bioconductor.org/p/110439/) in the support forum it is instead suggested to document the sampling and make it a user responsibility to call `set.seed`.  In our package, where for one call of `wasserstein.sc` the sampling function is called multiple times, this approach will not work.

Do you have any other suggestions on how to implement this feature? Is it acceptable to use `set.seed` if it is well documented in this case?

Thank you for your help.

Best,
Julian

**Comment by Liubuntu** [comment]
Hi @JulianFlesch , 

The `match.arg()` could actually detect the wrong input, so no need to have it as a option inside `switch`. https://github.com/goncalves-lab/waddR/blob/master/R/WassersteinTest.R#L360
 e.g., 
```{r}
fun <- function(type = c("ab", "cd", "ef")) { 
type <-  match.arg(type)
return(type)
}
fun()
## [1] "ab"
fun("random")
## Error in match.arg(type) : 'arg' should be one of “ab”, “cd”
```

Your first github link about using seed in R to pass to the Rcpp function seems working fine, this way, seems you can just add an argument e..g, `seed = integer()` whenever the `permnum` argument or else there for indication of calling that Rcpp function (with `sample` call). And in the script, add this check before each of the possible `sample` call: 

```
if (length(seed))  set.seed(seed)
```

I know somehow that Bioconductor would discourage the use of seed inside the script, but not quite in details. Unfortunately, bioconductor related links don't work due to some technical issues. I'll follow back here once I have more info. 

Best,
Qian

**Comment by Liubuntu** [comment]
Hi @JulianFlesch, 

After reading the support site post and your links of setting seed for Rcpp functions, I think you can have the `seed = NULL` argument in `wasserstein.test`, and pass if not NULL into the `.wassersteinTestSp` function, and finally Internally pass into the Rcpp function in `set.seed(seed)` (https://github.com/goncalves-lab/waddR/blob/master/src/wasserstein.cpp#L647 over here I guess). Remember to test if `is.null(seed)` before actually adding the `set.seed()`. Setting the default value as `NULL` is also important to avoid unexpected non-random permutation for users. 

Your explanation here for the use case https://github.com/Bioconductor/Contributions/issues/1218#issuecomment-537107633 will be kept for the record. 

Best,
Qian

**Comment by JulianFlesch** [comment]
Hi Qian,

Thank you for clarifying this. I will try to have the new version ready by next week for the second review. 

Best,
Julian

**Comment by JulianFlesch** [comment]
Hi @Liubuntu ,

Two changes were made in this version:

## 1. Reproducibility feature

There is now an extra argument called 'seed' in the `wasserstein.test` and `wasserstein.sc` functions. The seed value is passed on into the `.wassPermProcedure` function, where it is set in R just before calling the Rcpp function `permutate` ([this line](https://github.com/goncalves-lab/waddR/blob/master/R/WassersteinTest.R#L20) ). Rcpp seems to access the R RNG by default.

This example seems to highlight that this is indeed the desired behaviour:
```
library(waddR)

a <- c(1,2,3,4,1,2,3,4,1,2,3,1,1,1,2,2)
b <- a + 10
cond <- c(rep(1,length(a)), rep(2, length(b)))
m <- matrix(c(a,b,b,a), nrow=2, byrow=TRUE)

# in the following, the p-values are reproducible, while 
# 'rnorm' still returns a different sample each time
for (i in seq(1:5)) {
    print(wasserstein.sc(m, cond, method="OS", seed=2)[,"pval"])
    print(rnorm(10))
}
```

However, the use of `set.seed` causes BiocCheck to show a warning now.

## 2. Unnecessary argument checking removed

Removed default case in the switch statements in `wasserstein.sc` and `wasserstein.test`, because an equivalent error is already raised in `match.arg` (as you commented in a [previous comment](https://github.com/Bioconductor/Contributions/issues/1218#issuecomment-538124954)).


Thanks for your help!

Best, 
Julian

**Comment by Liubuntu** [comment]
Hi @mtmorgan , 

This package was accepted with one check warning for having `set.seed()` inside the script. Here is the explanation why the `set.seed()` was included in the package. https://github.com/Bioconductor/Contributions/issues/1218#issuecomment-537107633 

As the author explained, with one call of the `wassersteinTestSp` function, multiple calls of `sample()` inside the Rcpp function will be triggered. So if we leave the `set.seed()` call to user, it can not have reproducible result. Also, the `seed=NULL` was used as default for not including seed until users' necessary. 

Best,
Qian

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/JulianFlesch.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("waddR")`. The package 'landing page' will be created at

https://bioconductor.org/packages/waddR

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/

**Comment by LiNk-NY** [comment]
Hi @JulianFlesch 

There's no need to include the `seed` argument.
Please remove it from the code. `set.seed` will work fine outside the function. 

See the minimal example below:

``` r
library(waddR)

da <- rnorm(10, 10, 5)
db <- rnorm(10, 11, 3)

set.seed(100); head(permutations(c(da, db), 100))[, 1:5]
#>           [,1]      [,2]      [,3]      [,4]      [,5]
#> [1,] 19.750611 14.163236 19.750611  7.266167  7.266167
#> [2,]  6.277430  6.437196 12.022470 10.677698  6.890358
#> [3,]  7.266167  7.266167  6.437196 11.893725 12.022470
#> [4,] 15.216647 10.677698  7.587707  6.890358 11.893725
#> [5,] 11.126409 19.750611  7.266167 11.126409 15.216647
#> [6,]  7.241115 17.364220 11.126409 12.191647  1.733732

set.seed(100); head(permutations(c(da, db), 100))[, 1:5]
#>           [,1]      [,2]      [,3]      [,4]      [,5]
#> [1,] 19.750611 14.163236 19.750611  7.266167  7.266167
#> [2,]  6.277430  6.437196 12.022470 10.677698  6.890358
#> [3,]  7.266167  7.266167  6.437196 11.893725 12.022470
#> [4,] 15.216647 10.677698  7.587707  6.890358 11.893725
#> [5,] 11.126409 19.750611  7.266167 11.126409 15.216647
#> [6,]  7.241115 17.364220 11.126409 12.191647  1.733732

head(permutations(c(da, db), 100))[, 1:5]
#>           [,1]      [,2]      [,3]      [,4]      [,5]
#> [1,] 11.893725 17.364220 14.163236 11.893725  7.241115
#> [2,]  7.587707  7.587707  1.733732 10.677698 12.191647
#> [3,]  7.241115  6.643049  7.266167  3.902130  7.587707
#> [4,]  3.902130 19.750611 17.364220  7.587707  6.277430
#> [5,]  6.890358  6.890358  6.890358  7.266167  6.890358
#> [6,] 14.163236  6.277430 12.191647 12.022470 14.163236
```

<sup>Created on 2019-10-22 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

cc: @Liubuntu 

Best, 
Marcel

**Comment by mtmorgan** [comment]
Hmm, @LiNk-NY it seems though
```
> set.seed(124); wasserstein.sc(m, cond, method="OS")[,"pval"]; rnorm(3)
[1] 9.999e-05 9.999e-05
[1]  0.03832318 -0.76303016  0.21230614
> set.seed(124); wasserstein.sc(m, cond, method="OS")[,"pval"]; rnorm(3)
[1] 7.381147e-06 9.999000e-05
[1]  0.03832318 -0.76303016  0.21230614
```
And I think the reason is because of the use of parallel evaluation
```
> waddR:::.testWass
function (dat, condition, permnum, inclZero = TRUE, seed = NULL)
{
    if (!inclZero) {
        onegene <- function(x, dat, condition) {
            x1 <- dat[x, ][condition == unique(condition)[1]]
            x2 <- dat[x, ][condition == unique(condition)[2]]
            x1 <- (x1[x1 > 0])
            x2 <- (x2[x2 > 0])
            suppressWarnings(.wassersteinTestSp(x1, x2, permnum,
                seed))
        }
        wass.res <- bplapply(seq_len(nrow(dat)), onegene, condition = condition,
            dat = dat)
...
```
where eventually `.wassersteinTestSp()` calls 
```
> waddR:::.wassPermProcedure
function (x, y, permnum, seed = NULL)
{
    z <- c(x, y)
    if (!is.null(seed)) {
        set.seed(seed)
    }
    shuffle <- permutations(z, num_permutations = permnum)
...
```

Here's a simpler illustration using the parallel package
```
library(parallel)

f <- function(seed =NULL) {
    fun <- function(x) {
        if (!is.null(seed))
            set.seed(seed)
        rnorm(1)
    }
    unlist(mclapply(1:2, fun))
}
```
with reproducibility when the seed is set but not otherwise. 
```
> set.seed(123); f()
[1] -0.4378804  0.7307207
> set.seed(123); f()
[1] -0.1326210 -0.4025682
> set.seed(123); f(456)
[1] -1.343521 -1.343521
> set.seed(123); f(456)
[1] -1.343521 -1.343521
```
This is because, according to `?mcparallel` and working through the complicated description in the section 'Random Numbers',

> If the RNG has not yet been used [in the thread], the child will set a seed based on the time and process ID when it first uses the RNG

Notice too that setting the random number in the the forked process means that `rnorm()` returns the SAME NUMBER in each thread. In the original context the permutations are identical across threads, probably overstating the precision of the method.

This is very tricky, because the exact scheduling of jobs on the parallel processes is not entirely deterministic when there are more tasks than jobs. The solution seems to be to generate the random number seeds for each job 'up front' and pass that to the workers, and to generate the seeds in a properly independently manner, I think using `nextRNGStream()`. These streams would be passed to the worker and used to `set.seed()` on the worker, so somehow in violation of BIoc policy. But the take-home from the above discussion seems to be that the current use of set.seed() doesn't seem the best.

**Comment by JulianFlesch** [comment]
Hi @LiNk-NY,

As Martin's example shows, the current implementation with `set.seed` aims to make `wasserstein.sc` and `wasserstein.test` reproducible.

Hi @mtmorgan, 

Thanks for the explanations, we couldn't put our finger on why the seed doesn't seem to be used when set from outside `wasserstein.sc`.

> he take-home from the above discussion seems to be that the current use of set.seed() doesn't seem the best.

I will read up on the details of of `nextRNGStream` to try and have a better solution that enables reproducibility.

This is just to clarify what you are proposing in detail, please tell me if I am missing something:

> The solution seems to be to generate the random number seeds for each job 'up front' and pass that to the workers

You suggest to create an `nextRNGStream()` for each worker (`onegene`) before starting the first parallel jobs, and use `set.seed` inside the worker?

> scheduling of jobs on the parallel processes is not entirely deterministic

But these `nextRNGStrem()` instances must be created in association to genes (rows in the input; `wasserstein.test` is run on several genes in parallel), not the order of  job executions, because that might change? 

> In the original context the permutations are identical across threads, probably overstating the precision of the method.

This issue, we are aware of, but doesn't make the method less powerful. To avoid it, a user could select a vector of seeds for each gene, maybe like this:
```
set.seed(21)
seeds <- runif(n_genes)
```

Thank you for the feedback!

Best,
Julian

**Comment by mtmorgan** [comment]
I guess
```
set.seed(21)
seeds <- runif(n_genes)
```
is meant hypothetically; your current code couldn't use a vector of seeds. The problem with this, in a technical if not perhaps practical sense, is that there is precise mathematical relation between the elements of `seeds` so that any sequence of results based on seeds[1] will not be independent of seeds[2] -- the sequences will eventually synchronize. This is the purpose of `nextRNGStream()` -- to generate seeds that lead to independent streams.

Suppose one were working on a vector `genes`. One might implement the reproducible scenario by creating a task list where each element is a gene, and the seed associated with it
```
library(BiocParallel)
RNGkind("L'Ecuyer-CMRG")
set.seed(123)
seed <- .Random.seed
genes <- seq_len(10)
tasks <- vector("list", length(genes))
for (i in seq_along(genes)) {
    seed <- parallel::nextRNGStream(seed)
    tasks[[i]] <- list(gene=genes[[i]], seed=seed)
}
```
It would be used in `bplapply()` by passing each task to the worker, and setting the seed there...
```
t(simplify2array({
    bplapply(tasks, function(task) {
        RNGkind("L'Ecuyer-CMRG")
        .Random.seed <- task$seed
        c(task$gene, rnorm(1))
    })
}))
```
One could also
```
seed <- .Random.seed
seeds <- replicate(
    length(genes),
    { seed <<- parallel::nextRNGStream(seed) },
    simplify=FALSE
)

bpmapply(function(gene, seed) {
        RNGkind("L'Ecuyer-CMRG")
        .Random.seed <- seed
        c(gene, rnorm(1))
}, genes, seeds)
```

**Comment by mtmorgan** [comment]
Actually, I'll need to look at this a little more...
